syntax - sh.vim - Unify bashStatement, kshStatement and shStatement

bfoersterling opened the pull request:
## Changes

Following the discussion on the mailing list.

I have looked at all keywords and checked if they are built into `bash`,
`dash`, `ksh`, or `sh`, and if they are external commands.\
The external programs are still in `shStatement` but they are visually \
separated by a new vimscript statement in a new line.\
Unfortunatelly there are a lot of differences between builtin commands \
especially between the different Kornshell flavors and also Dash misses a lot \
of builtins.\
Maybe the external commands can also be moved to their own syntax group?\
But the highlighting will probably stay the same so not sure if this is useful.

Restructured and simplified the logic of the `kornshell` flavors.
Tried to unify statements whenever possible.

Highlighting for the `rpm` command was removed as being too distribution specific.

Cleanup: formatted with `gg=G` and removed dead code.

## Rerecorded Tests

- `testdir/dumps/sh_09_00.dump` (`let` is not a keyword in `dash`)
- `testdir/dumps/sh_ksh2020_05.dump` (`rename` is an ext. command that may be highlighted)
- `testdir/dumps/sh_ksh93u_05.dump` (`rename` is an ext. command that may be highlighted)
- `testdir/dumps/sh_ksh93v_05.dump` (`rename` is an ext. command that may be highlighted)
- `testdir/dumps/sh_ksh88_05.dump` (`rename` is an ext. command that may be highlighted)
- `testdir/dumps/sh_sundrous_03.dump` (`xgrep` should not be highlighted in `bash`)

```
$ let foo='bar'
dash: 1: let: not found
```

`rename` usually comes preinstalled through https://github.com/util-linux/util-linux/tree/master.

`xgrep` is neither a builtin in bash nor an external command that I could find.

bfoersterling committed:
refactored sh.vim

- unified bashStatement, kshStatement and shStatement as much as
  possible
- separated builtin commands from external programs
- cleaned up kornshell flavor logic

Signed-off-by: bfoersterling <bjoern.foersterling@gmail.com>

chrisbra commented on runtime/syntax/sh.vim:
```diff
@@ -463,76 +455,76 @@ syn match   shEscape	contained	'\%(^\)\@!\%(\\\\\)*\\.'	nextgroup=shComment
+    syn keyword bashSpecialVariables contained auto_resume BASH BASH_ALIASES BASH_ARGC BASH_ARGV BASH_CMDS BASH_COMMAND BASH_ENV BASH_EXECUTION_STRING BASH_LINENO BASHOPTS BASHPID BASH_REMATCH BASH_SOURCE BASH_SUBSHELL BASH_VERSINFO BASH_VERSION BASH_XTRACEFD CDPATH COLUMNS COMP_CWORD COMP_KEY COMP_LINE COMP_POINT COMPREPLY COMP_TYPE COMP_WORDBREAKS COMP_WORDS COPROC COPROC_PID DIRSTACK EMACS ENV ...
+    syn keyword bashStatement bind builtin caller compgen complete compopt declare dirs disown enable fg help history let logout mapfile popd pushd readarray shopt source suspend time
+    syn keyword bashStatement typeset nextgroup=shSetOption
+    syn keyword bashAdminStatement daemon reload restart start status stop
```
daemon is of little use, isn't it?

bfoersterling replied:
Right, I have removed it in the latest commit.

`reload`, `restart`, `start`, `status`, and `stop` are not that useful either in my opinion, but maybe someone wants these for `sys5init` scripts.\
And I don't think these script have to be in `bash`, so it should rather be named `shAdminStatement` and put in global scope.

dkearns commented on runtime/syntax/sh.vim:
```diff
@@ -797,49 +786,29 @@ syn region shParen matchgroup=shArithRegion start='\$\@!(\%(\ze[^(]\|$\)' end=')
 " ===================
-syn keyword shStatement break cd chdir continue eval exec exit kill newgrp pwd read readonly return shift test trap ulimit umask wait
+" builtins:
+syn keyword shStatement alias bg break cd continue command eval exec exit export fc getopts hash jobs local printf read readonly return shift times trap type ulimit umask unalias wait
```
I haven't tried but this should break the `shAlias` match as the `syn-keyword` will have priority.

bfoersterling replied:
Yes, it actually broke it for some reason.

I changed it back and added the test `runtime/syntax/testdir/input/sh_bash_alias.sh`.\
Maybe you can check if it is correct now and if you have an edge case in mind add a test for it.

Thanks for the review.

bfoersterling committed:
alias and daemon keywords

- fixed alias syntax highlighting
- added test for bash alias syntax highlighting
- removed daemon keyword

Signed-off-by: bfoersterling <bjoern.foersterling@gmail.com>
